Replace FocusTrap with TapRegionSurface

gspencergoog opened the pull request:
A PR to address https://github.com/flutter/flutter/issues/86972

The code derives largely from the suggestion in [this comment](https://github.com/flutter/flutter/issues/86972#issuecomment-1065788643), as well as some explorations I had while working on the `MenuBar` widget. It differs from Venkat's code in that there is also a grouping mechanism for the regions, so that multiple regions can act as one, allowing things like taps on the text field toolbars to not unfocus the text field.

The way `TapRegionSurface` is supposed to work is that it wraps the whole app area, and triggers any `onTapOutside` callbacks for regions or groups of regions when a tap occurs.

This is a shift from how `FocusTrap` works, in that widgets that want to have the focus disappear will need to do some setup instead of getting it by default.  Since `TextField` is the main widget that wants to do this, that shouldn't really be too hard. I added an `onTapOutside` and `tapRegionGroupId` attributes to `TextField` that if unset, will do what `FocusTrap` used to do (only lose focus on desktop and web platforms). But you can also set it if you want to override that behavior, or disable it, and place the `TextField` inside of a `TapRegion` group that includes, for instance, value spinners or a "copy to clipboard" button so that they won't affect the focus.

Usage looks something like this:
```dart
Widget build(BuildContext context) {
  return Column(
    children: <Widget>[
      const Text('Outside Surface'),  // Taps on this would not be considered.
      TapRegionSurface( // This would normally be at the top of the app.
        child: Row(
          children: <Widget>[
            const Text('Outside'),
            TapRegion(
              onTapOutside: () {
                clickedOutside.add('No Group');
              },
              child: const Text('No Group'),
            ),
            TapRegion(
              groupId: 1,
              onTapOutside: () {
                clickedOutside.add('Group 1 A');
              },
              child: const Text('Group 1 A'),
            ),
            TapRegion(
              groupId: 1,
              onTapOutside: () {
                clickedOutside.add('Group 1 B');
              },
              child: const Text('Group 1 B'),
            ),
          ],
        ),
      ),
    ],
  );
}
```

cc @justinmc @venkatd @jpnurmi

gspencergoog commented:
In looking at the test failures, it's clear that we'd have to figure something out for the cut/copy/paste toolbar so that clicking on it doesn't cause the field to be unfocused.

gspencergoog commented:
OK, after a few false starts, this is pretty much exactly what @venkatd had, with the addition of groups.

I've also integrated groups into the selection dialogs so that taps on the overlay dialogs don't unfocus the text field.

gspencergoog committed:
Replace FocusTrap with TapRegionSurface

gspencergoog commented:
There's still some documentation that needs to be written for this, but does this seem like a reasonable replacement for `FocusTrap`?

gspencergoog commented:
And probably some name refinement.  I went with `TapRegion` instead of `TapOutsideDetector`, since I was hoping to also allow non-gesture-arena "onTapInside" callbacks. Those are terribly useful, but it might be interesting to know that there was a tap on something inside your "group", and not just outside your group.

gspencergoog committed:
Add onTapInside

gspencergoog committed:
Add more tests, fix border detection.

gspencergoog committed:
Fix tests

gspencergoog committed:
Fix comment

gspencergoog committed:
Make TapRegion enable/disable instead of add/don't add

gspencergoog committed:
Fix test

gspencergoog committed:
Revert unnecessary test changes

gspencergoog committed:
Reapply some changes for web tests to pass

gspencergoog commented:
Okay, I this now has tests and minimal changes to existing tests.  The changes that are there are all related to web (without the change, they only fail on web), mainly because web does its own rendering of the toolbar and handles.

justinmc reviewed (commented):
I think this is a great solution.  I've been pretty invested in getting a solution to this problem right, and this PR seems to solve all of the use cases I had for it.

### Text formatting buttons

One big problem I had with FocusTrap was the inability to create a button that doesn't steal focus from a text field, like a "bold" button in a text editor.  I tried it and was easily able to achieve this with your PR :+1:.  Do you think it's worthwhile to include a code example of how to do this in the docs somewhere?  Just something like this:

```dart
        TapRegion(
          groupId: groupId,
          child: TextButton(
            onPressed: () {
              print('pressed bold');
            },
            child: const Text('Bold'),
          ),
        ),
        const TextField(
          tapRegionGroupId: groupId,
        ),
```

### Virtual keyboard
This was another thing I thought should be possible (brought up by @jpnurmi, see [this comment](https://github.com/flutter/flutter/issues/86972#issuecomment-902155543)).  With this PR it is possible but not ideal.  I guess you would have to use the same groupId with your virtual keyboard and all text fields.  Maybe that's acceptable?  You also couldn't create a virtual keyboard Pub package that just worked out of the box, it would also have to bundle in its own text fields or something.

Maybe this is all fine and there's nothing to improve here.

### Gesture arena
Despite not integrating with the gesture arena and not supporting gestures other than taps, this solution seems to work great for all of my use cases.  However I can't help but wonder if there isn't a more generic solution that would work with the gesture arena for any gesture Flutter supports.  I think it's probably just something to monitor as people get their hands on this since it would be a huge change.

### Misc.

I modified your code example above slightly to test this locally, here's my code for the record:

<details>

<summary>Demo code</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
Widget build(BuildContext context) {
  return Scaffold(
    body: Column(
      children: <Widget>[
        const Text('Outside Surface'),  // Taps on this would not be considered.
        const SizedBox(height: 200.0),
        TapRegionSurface( // This would normally be at the top of the app.
          child: Container(
            color: Colors.red.withOpacity(0.2),
            padding: const EdgeInsets.all(32.0),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: <Widget>[
                const Text('Outside'),
                Container(
                  color: Colors.green.withOpacity(0.2),
                  child: TapRegion(
                    onTapOutside: () {
                      print('onTapOutside of No Group');
                    },
                    child: const Text('No Group'),
                  ),
                ),
                Container(
                  color: Colors.blue.withOpacity(0.2),
                  child: TapRegion(
                    groupId: 1,
                    onTapOutside: () {
                      print('onTapOutside of Group 1 A');
                    },
                    child: const Text('Group 1 A'),
                  ),
                ),
                Container(
                  color: Colors.purple.withOpacity(0.2),
                  child: TapRegion(
                    groupId: 1,
                    onTapOutside: () {
                      print('onTapOutside of Group 1 B');
                    },
                    child: const Text('Group 1 B'),
                  ),
                ),
              ],
            ),
          ),
        ),
      ],
    ),
  );
}
}
```

</details>

gspencergoog commented:
> Do you think it's worthwhile to include a code example of how to do this in the docs somewhere? 

Yes, I'll add one with that.

gspencergoog committed:
Fix some tests so they don't need modification

gspencergoog commented:
OK, in trying to remove the last few lines that disable the tap detection in some tests, I realized that the reason they're failing with it disabled is that tapping on another text field is causing the focus to be lost in the first text field due to the TapRegion and gained in the second text field again once it's clicked. This causes the keyboard to "bounce" because the click on a second text field causes the keyboard to be hidden and then shown again.

One solution to this is to make all text fields (and their decorations and toolbars, etc.) share the same group ID.

Making them all have the same ID "works", but it makes the API less appealing, because if i let people override the ID, they'll be surprised by the keyboard bouncing behavior, and if they want to make their own text field components that don't cause loss of focus, they'll have to use a const `EditableText.groupId` as the ID (they can't make their own).  It means it's only useful to make your own ID if the group you want to have isn't a text field. Which means I might as well have a widget called `TextFieldTapRegion` that you wrap anything you want to be treated as part of a text field with, so they don't have to know about the magic ID, and then just not allow them to set the ID on text fields.

This seems to only be a problem on the iOS browser, so maybe we can do something to prevent the bouncing (probably something we already do on Android)? It does seem like in general, we would want hiding the keyboard to wait for 100ms or something before actually hiding it, in order to know if something else is just going to immediately show it again.

gspencergoog committed:
Remove tapRegionGroupId, make all text fields use the same group

gspencergoog committed:
Update macro name

jonahwilliams reviewed (commented):
This is sort of a nit, but I've always thought of a tap as referring to a primary pointer down but only with touch input - whereas click would be primary pointer down with mouse input. I'm not sure if we make a distrinction elsewhere in the framework, or if it is worth making one here

gspencergoog committed:
Fix remaining web tests, hopefully

gspencergoog commented:
Yeah, I agree, but since it can be either one in this case, what term would you use?  I used tap just because it's more common in our APIs to have an "onTap" than an "onClick".

gspencergoog commented:
We could invent a new term, 'onDepress' or something. :-) (but `DepressRegion`?)

(and there are actually no instances of "onClick" anywhere in the framework).

jonahwilliams commented:
I would say the prevalence of ontap vs onclick is definitely a testament to our mobile first heritage. No ideas here

gspencergoog committed:
Fix CupertinoTabBar so tapping on a tab doesn't lose focus on contents

gspencergoog committed:
Cleanup

gspencergoog committed:
Add text field spinner example

gspencergoog commented:
OK, @jonahwilliams and @justinmc, this is ready for another look.  I've added `TextFieldTapRegion` as a thin wrapper around `TapRegion`, made all text fields part of the same group (even when not focused) and added an example that adds a spinner to a text field (without unfocusing the text field when you click them).

justinmc commented on examples/api/lib/widgets/tap_region/text_field_tap_region.0.dart:
```diff
@@ -0,0 +1,250 @@
+      ),
+    );
+  }
+}
```
Very thorough example, thanks for writing it!  And testing it.

justinmc commented on examples/api/test/widgets/tap_region/text_field_tap_region.0_test.dart:
```diff
@@ -0,0 +1,90 @@
+    expect(find.byIcon(Icons.add), findsOneWidget);
+    expect(find.byIcon(Icons.remove), findsOneWidget);
+  });
+  testWidgets('tapping increment/decrement works', (WidgetTester tester) async {
```
Nit: newlines in between the testWidgets if that's a standard Flutter thing?

justinmc commented on examples/api/test/widgets/tap_region/text_field_tap_region.0_test.dart:
```diff
@@ -0,0 +1,90 @@
+    expect(
+      getFieldValue(tester).selection,
+      equals(const TextSelection(baseOffset: 0, extentOffset: 3)),
+    );
```
Is it possible to also test that the field has focus still?

justinmc commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -3452,102 +3488,126 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
+  // The default behavior if widget.onTapOutside is null.
+  void _handleDefaultTapOutside() {
+    /// The focus dropping behavior is only present on desktop platforms
+    /// and mobile browsers.
```
Maybe this PR is not the place to worry about this, but does the behavior change for mobile devices with mice or desktop devices with touchscreens?  Relevant issue: https://github.com/flutter/flutter/issues/106586

justinmc commented on packages/flutter/lib/src/widgets/tap_region.dart:
```diff
@@ -0,0 +1,534 @@
+///
+///  * [TapRegion], the widget that this widget uses to add widgets to the group
+///    of text fields.
+class TextFieldTapRegion extends TapRegion {
```
+1 This seems like a clean way to do this while allowing the user to put their own stuff in the same group too.

justinmc reviewed (approved):
LGTM :+1: 

I'm excited that this will make the "virtual keyboard" use case easy to do as well.

gspencergoog committed:
Review changes

Eerey commented:
Will this also work for a desktop environment (windows) when using a virtual keyboard in combination with a WebView inside flutter?

Use Case: Fill in forms (web) inside a WebView component with a virtual keyboard (when there is no physical keyboard present). 

We'd love to see this work, since the windows OSK is... not visually appealing.

gspencergoog commented:
Hi @Eerey 

> Will this also work for a desktop environment (windows) when using a virtual keyboard in combination with a WebView inside flutter?

To be clear: you're talking about the case of building your own virtual keyboard, right?  Not a system built-in keyboard?

You should be able to do that, but you might need to wrap the web view in a `TapRegion` widget, and do the same to your virtual keyboard (using the same `groupId` for both) to keep the web view from losing focus when you tap on the keyboard buttons. If you're mixing Flutter text fields and web view text fields in the same UI, you might need to use a `TextFieldTapRegion` around them instead, to let Flutter think that the web view and your keyboard are both part of the text field group.

With this change, building a virtual keyboard should be a lot easier, however.

Eerey commented:
Good evening @gspencergoog 
> With this change, building a virtual keyboard should be a lot easier, however.

That sounds **amazing**. I am very much looking forward to it.

Eerey commented:
Hi, where can we observe if this feature is currently present? When will this be available in the master channel? :)

gspencergoog commented:
@Eerey It is currently present in the master and beta channels.  It will be in stable with the next stable release.

Just for future reference, the master channel is the first place where things are added, so something new is always available in the master channel. There is only a delay for things going into the more stable channels (beta and stable). See https://github.com/flutter/flutter/wiki/Flutter-build-release-channels for more information.

Jjagg commented:
@gspencergoog Should this not have gone through the [breaking change policy](https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes) by including a design doc and migration guide for `FocusTrap`/`FocusTrapArea`?
I had to dig a little to figure out how to migrate.

gspencergoog commented:
@Jjagg  I'm sorry it was hard to find.  Hopefully the migration wasn't too hard, and that it meets your needs better than `FocusTrap` did. And no, this change didn't break any of our contributed tests, so it wasn't considered a breaking change for the purposes of the policy.

Jjagg commented:
@gspencergoog It wasn't too bad and this is a way more flexible solution, so I'm very happy with this change :)
Thank you for your work on it.

Ah sorry, I thought public API changes fell under the policy. I'll have to remember to submit some stuff to the contributed tests sometimes.
